From 61ee97e915da51873700ec30ac46888cdc732c3a Mon Sep 17 00:00:00 2001
From: Arjan van de Ven <arjan@linux.intel.com>
Date: Sat, 11 Jun 2016 22:12:51 +0000
Subject: [PATCH 123/129] printk end of boot

---
 init/main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/main.c b/init/main.c
index 92cdd9cef1b3..3e65e6a3fe13 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1078,6 +1078,7 @@ static int __ref kernel_init(void *unused)
 		       ramdisk_execute_command, ret);
 	}
 
+	printk(KERN_ERR "Executing init\n");
 	/*
 	 * We try each of these until one succeeds.
 	 *
-- 
2.18.0

