From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Arjan van de Ven <arjan@linux.intel.com>
Date: Sat, 11 Jun 2016 22:12:51 +0000
Subject: [PATCH] printk end of boot

---
 init/main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/main.c b/init/main.c
index e6836a9400d5..ae52073b12d2 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1470,6 +1470,7 @@ static int __ref kernel_init(void *unused)
 		       ramdisk_execute_command, ret);
 	}
 
+	printk(KERN_ERR "Executing init\n");
 	/*
 	 * We try each of these until one succeeds.
 	 *
-- 
https://clearlinux.org

