--- linux-4.6/crypto/rsa.c~	2016-05-15 22:43:13.000000000 +0000
+++ linux-4.6/crypto/rsa.c	2016-05-26 00:02:54.646498761 +0000
@@ -318,10 +318,13 @@
 {
 	int err;
 
+	printk("ONE\n");
 	err = crypto_register_akcipher(&rsa);
 	if (err)
 		return err;
 
+	printk("TWO\n");
+
 	err = crypto_register_template(&rsa_pkcs1pad_tmpl);
 	if (err) {
 		crypto_unregister_akcipher(&rsa);
--- linux-4.6/crypto/algapi.c~	2016-05-15 22:43:13.000000000 +0000
+++ linux-4.6/crypto/algapi.c	2016-05-26 00:14:16.157803498 +0000
@@ -367,7 +367,7 @@
 	if (IS_ERR(larval))
 		return PTR_ERR(larval);
 
-	crypto_wait_for_test(larval);
+//	crypto_wait_for_test(larval);
 	return 0;
 }
 EXPORT_SYMBOL_GPL(crypto_register_alg);
