--- linux-4.7/init/main.c.org	2016-08-04 17:57:14.121653178 +0000
+++ linux-4.7/init/main.c	2016-08-04 17:57:56.366072529 +0000
@@ -929,6 +929,12 @@
 }
 #endif
 
+static void async_free_initmem(void *data, async_cookie_t cookie)
+{
+	free_initmem();
+	mark_readonly();
+}
+
 static int __ref kernel_init(void *unused)
 {
 	int ret;
@@ -936,8 +942,7 @@
 	kernel_init_freeable();
 	/* need to finish all async __init code before freeing the memory */
 	async_synchronize_full();
-	free_initmem();
-	mark_readonly();
+	async_schedule(async_free_initmem, NULL);
 	system_state = SYSTEM_RUNNING;
 	numa_default_policy();
 
