--- linux-4.7/include/linux/mm.h~	2016-07-24 19:23:50.000000000 +0000
+++ linux-4.7/include/linux/mm.h	2016-08-02 16:00:02.825803647 +0000
@@ -2071,8 +2071,8 @@
 void task_dirty_inc(struct task_struct *tsk);
 
 /* readahead.c */
-#define VM_MAX_READAHEAD	128	/* kbytes */
-#define VM_MIN_READAHEAD	16	/* kbytes (includes current page) */
+#define VM_MAX_READAHEAD	512	/* kbytes */
+#define VM_MIN_READAHEAD	32	/* kbytes (includes current page) */
 
 int force_page_cache_readahead(struct address_space *mapping, struct file *filp,
 			pgoff_t offset, unsigned long nr_to_read);
