From 94a9049ff05bd8a0a882b692014c8bec1c87f8db Mon Sep 17 00:00:00 2001
From: Arjan van de Ven <arjan@linux.intel.com>
Date: Thu, 4 Aug 2016 18:07:50 +0000
Subject: [PATCH 128/129] free initmem asynchronously

---
 init/main.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/init/main.c b/init/main.c
index 1f3f3db51de9..e1133f95c0ef 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1054,6 +1054,12 @@ static inline void mark_readonly(void)
 }
 #endif
 
+static void async_free_initmem(void *data, async_cookie_t cookie)
+{
+	free_initmem();
+	mark_readonly();
+}
+
 static int __ref kernel_init(void *unused)
 {
 	int ret;
@@ -1063,8 +1069,7 @@ static int __ref kernel_init(void *unused)
 	async_synchronize_full();
 	ftrace_free_init_mem();
 	jump_label_invalidate_initmem();
-	free_initmem();
-	mark_readonly();
+	async_schedule(async_free_initmem, NULL);
 	system_state = SYSTEM_RUNNING;
 	numa_default_policy();
 
-- 
2.19.0

